### ISSUE TITLE
implicit-dynamic: fals, anonymous functions, and explicitly dynamic types require gratuitous verbosity in code

### ISSUE THREAD

**Issue by kevmoo** [dart-lang/language#993]
Supporting the generation of code that's compliant with `implicit-dynamic: false` is is a BIG request for json_serializable

https://github.com/dart-lang/json_serializable/issues/557

The VAST majority of the cases in generate code map to something like this

image.png
![image](https://user-images.githubusercontent.com/17034/83188304-55299e80-a0e4-11ea-9302-16a80b61e601.png)

 error: Missing parameter type for 'e'. (implicit_dynamic_parameter at [json_serializable] test/kitchen_sink/kitchen_sink.g_exclude_null__non_nullable.g.dart:117)

`e` in this case maps back to something explicitly dynamic. Adding explicit `dynamic` here seems silly and redundant.

Is there some deeper logic here I'm missing? Could we update this error?

**Comment by rrousselGit** [comment]
Not sure about the reasoning, but there are other similar situations, such as:

```dart
void function(List<dynamic> list) {}

function([]); // error
```

**Comment by leafpetersen** [comment]
Agreed that `--no-implicit-dynamic` is too coarse.  That's why @srawlins and I were pushing on `--strict-inference` as a replacement.  My plan, once null safety has shipped, is to finish off the suite of `--strict` flags, and try to get widespread adoption.  Specifically to the examples above, if you use strict inference:

```
analyzer:
  language:
    strict-inference: true
```

then you get the following:

```dart
void function(List<dynamic> list) {}

void foo(void Function(dynamic x)) {}
void main() {
  function([]); // ok
  var f = (e) {}; // error
  foo((e) {}); // ok
}
```

which I believe is what you want in the cases above.

The remaining pain point, which I hope to address after null safety is that `--strict-raw-types` is, in my opinion, too noisy.  I have some ideas about how to make it work well, but they have to wait until after the null safety release.

**Comment by rrousselGit** [comment]
Where can we learn more about those --strict flags?

They don't seem to be documented https://dart.dev/guides/language/analysis-options

**Comment by leafpetersen** [comment]
The current flags are documented [here](https://github.com/dart-lang/language/blob/master/resources/type-system/strict-inference.md) and [here](https://github.com/dart-lang/language/blob/master/resources/type-system/strict-raw-types.md).

We've ended up not having the resources to push hard on these so far this year, but I'm hopeful that I can turn some attention to them after the push to null safety is over.
